File: Python/email-api.py
#!/usr/bin/env python
# Credentials are read in from a configuration file in the same directory. 
# The file is named config.cfg and takes the following format:
#
# [Credentials]
# customer: [customer]
# username: [username]
# password: [password]
#

import sys, json, httplib2, ConfigParser
from urllib import urlencode

# Create a new instance of the http2lib module
http = httplib2.Http()
# Force errors to be returned as normal responses
http.force_exception_to_status_code = True

# Create a new instance of the ConfigParser module
config = ConfigParser.ConfigParser()
try:
    # Read in the credentials from config.cfg
    config.read('config.cfg')
    apikey = config.get('Credentials', 'apikey', 'none')
except Exception, ex:
    # Exit if config file parsing failed
    print str(ex)
    sys.exit("Error reading config.cfg")

#####POST#####
# Create or modify an account
def createEditAccount():
    # Specify API resource URI and parameters
    uri = 'https://emailapi.dynect.net/rest/json/accounts'
    params = {
        'apikey': apikey,
        'username': 'username',
        'password': 'password',
        'companyname': 'Example, Inc.',
        'phone': '555-555-5555'
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Make the web request
    response, content = http.request(uri, 'POST', contents, headers=header)
    result = json.loads(content)

    # Print result on success
    if result["response"]["status"] == 200:
        print "Account created or modified successfully."
        return 1
    # Throw an error message if request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

#####POST#####
# Add to the list of approved recipients
def addSenders():
    # Specify API resource URI and parameters
    uri = 'http://emailapi.dynect.net/rest/json/senders'
    params = {
        'apikey': apikey,	
        'emailaddress':'example@example.com'
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Make the web request
    response, content = http.request(uri, 'POST', contents, headers=header)
    result = json.loads(content)

    # Print result on success
    if result["response"]["status"] == 200:
        print "Sender added successfully."
        return 1
    # Throw an error message if request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

#####POST#####
# Add to the list of approved recipients
def addRecipients():
    # Specify API resource URI and parameters
    uri = 'http://emailapi.dynect.net/rest/json/recipients/activate'
    params = {
        'apikey': apikey,	
        'emailaddress': 'example@example.com'
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Make the web request
    response, content = http.request(uri, 'POST', contents, headers=header)
    result = json.loads(content)

    # Print result on success
    if result["response"]["status"] == 200:
        print "Recipient added successfully."
        return 1
    # Throw an error message if request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

#####GET#####
# Return number of successfully delivered emails
def reportDelivered():
    # Specify parameters for API request
    params = {
        'apikey': apikey,
        'startdate': '2013-12-01',
        'enddate': '2014-01-31'
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Append HTTP query form info to the URI
    uri = 'http://emailapi.dynect.net/rest/json/reports/sent/count?' + contents

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Make the web request
    response, content = http.request(uri, 'GET', contents, headers=header)
    result = json.loads(content)

    # Print result on success
    if result["response"]["status"] == 200:
        print "Messages sent successfully: " + str(result["decode"]["response"]["data"]["count"])
        return 1
    # Throw an error message if request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

#####GET#####
# Return a report of bounced emails
def reportBounced(startIndex):
    # Specify parameters for API request
    params = {
        'apikey': apikey,
        'startdate': '2013-12-01',
        'enddate': '2014-01-31',
        'startindex': startIndex
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Append HTTP query form info to the URI
    uri = 'http://emailapi.dynect.net/rest/json/reports/bounces?' + contents

    # Make the web request
    response, content = http.request(uri, 'POST', contents, headers=header)
    result = json.loads(content)

    count = 0
    # If the API did not respond with an error
    if result["response"]["status"] == 200:
        # Cycle through each bounce record
        for bounce in result["response"]["data"]["bounces"]:
            # Print up to 500 bounce records on success
            if (count < 500):
                print "Bounce type: " + bounce["bouncetype"]
                print "Bounce rule: " + bounce["bouncerule"]
                print "Sender address: " + bounce["emailaddress"]
                print "Bounce time: " + bounce["bouncetime"] + "\n"
            # If 500 records have already been returned
            # Call the function again with a start index incremented by 500
            else:
                reportBounced(startIndex + 500);
                return

    # Throw an error message if API request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

#####POST#####
# Send email with the Dyn Message Manager API
def sendEmail():
    # Specify API resource URI and parameters
    uri = 'http://emailapi.dynect.net/rest/json/send'
    params = {
        'apikey': apikey,	
        'from': 'example@example.com',
        'to': 'example2@example.com',
        'subject': 'Test',
        'bodytext': 'Test test test',
        'bodyhtml': 'Test test test'
    }
    # Set the header for the web request
    header = {'content-type': 'application/x-www-form-urlencoded'}

    # Convert parameters into an HTTP query form, delimeted by &
    # Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
    contents = urlencode(params)

    # Make the web request
    response, content = http.request(uri, 'POST', contents, headers=header)
    result = json.loads(content)

    # Print result on success
    if result["response"]["status"] == 200:
        print "Message sent successfully."
        return 1
    # Throw an error message if request is not successful
    else:
        print "API Error:"
        print "Status: " + str(result["response"]["status"])
        print "Message: " + str(result["response"]["message"])
        return 0

File: Perl/email-api.pl
#!/usr/bin/perl
# Credentials are read in from a configuration file in the same directory. 
# The file is named config.cfg and takes the following format:
#
# customer: [customer]
# username: [username]
# password: [password]
#

use warnings;
use strict;
use Config::Simple;
use LWP::UserAgent;
use URI;
use URI::QueryParam;
use JSON;

# Create config reader
my $config = new Config::Simple();
# Read configuration file (can fail)
$config->read('config.cfg') or die $config->error();
# Dump config variables into hash for later use
my %configOptions = $config->vars();

# Read in API key from config.cfg
my $apiKey = $configOptions{'apikey'} or do {
	print "API key required in config.cfg for API login\n";
	exit;
};
# Optioanl: read in username from config.cfg
my $username = $configOptions{'username'};
# Optional: read in password from config.cfg
my $password = $configOptions{'password'};

#####POST#####
# Create or modify an account
sub createEditAccount {
	# Specify API resource URI and parameters
	my $uri = 'https://emailapi.dynect.net/rest/json/accounts';
	my %params = (
		'apikey' => $apiKey,
		'username' => $username,
		'password' => $password,
		'companyname' => 'Test Inc.',
		'phone' => '555-555-5555'
	);
	
	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
	
	# Set up the HTTP request
	my $request = HTTP::Request->new(POST => $uri);
	$request->header('content-type' => 'application/x-www-form-urlencoded');
	$request->content($contents);
	
	# Make the web request
	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);
	
	# Print results
	if ($decode->{'response'}->{'status'} == 200) {
		print "Account succesfully created or modified.";
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}

#####POST#####
# Add to the list of approved senders
sub addSenders {
	# Specify API resource URI and parameters	
	my $uri = 'http://emailapi.dynect.net/rest/json/senders';
	my %params = (
		'apikey' => $apiKey,	
		'emailaddress' => 'example@example.com'
	);
	
	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
	
	# Set up the HTTP request
	my $request = HTTP::Request->new(POST => $uri);
	$request->header('content-type' => 'application/x-www-form-urlencoded');
	$request->content($contents);

	# Make the web request
	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);
	
	# Print results
	if ($decode->{'response'}->{'status'} == '200') {
		print "Sender added succesfully.\n";
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}

#####POST#####
# Add to the list of approved recipients
sub addRecipients {
	# Specify API resource URI and parameters
	my $uri = 'http://emailapi.dynect.net/rest/json/recipients/activate';
	my %params = (
		'apikey' => $apiKey,	
		'emailaddress' => 'example@example.com'
	);
	
	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
	
	# Set up the HTTP request
	my $request = HTTP::Request->new(POST => $uri);
	$request->header('content-type' => 'application/x-www-form-urlencoded');
	$request->content($contents);

	# Make the web request
	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);
	
	# Print results
	if ($decode->{'response'}->{'status'} == 200) {
		print "Recipient added succesfully.\n";
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}

#####GET#####
# Return number of successfully delivered emails
sub reportDelivered {
	#####Need to add a starttime and endtime#####
 	# Specify parameters for API request
 	my %params = (
 		'apikey' => $apiKey,
 		'starttime' => '2013-12-01',
 		'endtime' => '2014-01-31'
 	);
 	
 	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
 	
 	# Append HTTP query form info to the URI
	my $uri = 'http://emailapi.dynect.net/rest/json/reports/sent/count?' . $contents;
 
	# Set up the HTTP request
	my $request = HTTP::Request->new(GET => $uri);
	$request->header('content-type' => 'application/json');
 	
 	# Make the web request
 	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);
	
	# Print results
	if ($decode->{'response'}->{'status'} == 200) {
		print "Messages delivered successfully: " . $decode->{'response'}->{'data'}->{'count'} . "\n";
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}

#####GET#####
# Return a report of bounced emails
# Expects start index argument
# First record can be found at 0
sub reportBounced {
	my $startIndex = @_;
	# Specify parameters for API request
 	my %params = (
 		'apikey' => $apiKey,
 		'starttime' => '2013-11-01',
 		'endtime' => '2014-01-31',
 		'startindex' => $startIndex
 	);
 	
 	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
 	
 	# Append HTTP query form info to the URI
	my $uri = 'http://emailapi.dynect.net/rest/json/reports/bounces?' . $contents;
	
	# Set up the HTTP request
	my $request = HTTP::Request->new(GET => $uri);
	$request->header('content-type' => 'application/json');
 	
 	# Make the web request
 	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);

	my $count = 0;
	if ($decode->{'response'}->{'status'} == 200) {
		foreach my $bounce ($decode->{'response'}->{'data'}->{'bounces'}[0]) {
			# Print up to 500 bounce records on success
			if ($count < 500) {
				print "Bounce type: " . $bounce->{'bouncetype'} . "\n";
				print "Bounce rule: " . $bounce->{'bouncerule'} . "\n";
				print "Sender address: " . $bounce->{'emailaddress'} . "\n";
				print "Bounce time: " . $bounce->{'bouncetime'} . "\n";
			}
			else {
				reportBounced($startIndex + 500);
				return;
			}
		}
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}

#####POST#####
# Send email with the Dyn Message Manager API
sub sendEmail {
	# Specify API resource URI and parameters
	my $uri = 'http://emailapi.dynect.net/rest/json/send';
	my %params = (
		'apikey' => $apiKey,	
		'from' => 'example@example.com',
		'to' => 'example2@example.com',
		'subject' => 'Test',
		'bodytext' => 'Test test test test.'
	);
	
	# Convert parameters into an HTTP query form, delimeted by &
	# Example: apikey=dnhtir34ty98uyghdfknfh3&emailaddress=example%40example.com
	my $uriQuery = URI->new(" ", "http");
	$uriQuery->query_form_hash(\%params);
	my $contents = $uriQuery->query;
	
	# Set up the HTTP request
	my $request = HTTP::Request->new(POST => $uri);
	$request->header('content-type' => 'application/x-www-form-urlencoded');
	$request->content($contents);

	# Make the web request
	my $lwp = LWP::UserAgent->new;
	my $response = $lwp->request($request);
	my $decode = decode_json($response->content);
	
	# Print results
	if ($decode->{'response'}->{'status'} == 200) {
		print "Mail sent successfully.\n";
	}
	# Throw an error message if request is not successful
	else {
		print "API Error:\n";
		print "Status: " . $decode->{'response'}->{'status'} . "\n";
		print "Message: " . $decode->{'response'}->{'message'} . "\n";
	}
	return;
}
